crypto/s2n_certificate.c

```diff
@@ -238,8 +275,54 @@ int s2n_send_cert_chain(struct s2n_stuffer *out, struct s2n_cert_chain *chain)
     return 0;
 }
 
-int s2n_send_empty_cert_chain(struct s2n_stuffer *out) {
+int s2n_send_empty_cert_chain(struct s2n_stuffer *out)
+{
     notnull_check(out);
     GUARD(s2n_stuffer_write_uint24(out, 0));
     return 0;
 }
+
+static int s2n_does_cert_san_match_hostname(struct s2n_cert_chain_and_key *cert, const char *hostname)
+{
+    GENERAL_NAMES *san_names = cert->san_names;
+    if (san_names == NULL) {
+        return 0;
+    }
+
+    const int hostname_len = strnlen(hostname, S2N_MAX_SERVER_NAME);
+    const int num_san_names = sk_GENERAL_NAME_num(san_names);
+    for (int i = 0; i < num_san_names; i++) {
+        GENERAL_NAME *san_name = sk_GENERAL_NAME_value(san_names, i);
+        if (!san_name) {
+            continue;
+        }
+
+        /* we only care about DNS entries */
+        if (san_name->type == GEN_DNS) {
+            unsigned char *san_str = NULL;
+            const int san_str_len = ASN1_STRING_to_UTF8(&san_str, san_name->d.dNSName);
```

Comment:
I was looking at curl's source code and they don't convert the GENERAL_NAME string to UTF. They do a direct comparison on the ASN1 SAN string. I'll investigate more and remove this UTF8 conversion if possible. We can avoid a memory allocation.

Reply:
Peeked a bit at:
* https://github.com/curl/curl/blob/3f5da4e59a556fc68272a9857a38dd75234d0c04/lib/vtls/openssl.c#L1562-L1577
* https://github.com/openssl/openssl/blob/4286ca47c7fac2b5a6208b377967e0f095e9abc5/crypto/x509v3/v3_utl.c#L836-L845 https://github.com/openssl/openssl/blob/4286ca47c7fac2b5a6208b377967e0f095e9abc5/crypto/x509v3/v3_utl.c#L793-L794 

Since the only valid type for dNSName is V_ASN1_IA5STRING , we don't need to do a decoding here. Just a case insensitive comparison.

Reply:
added this in 
140854d